File: src/XOPE_UI.Spy/ServerType/IMessageWithResponse.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace XOPE_UI.Spy.ServerType
{
    public abstract class IMessageWithResponse : IMessage
    {
        public event EventHandler<JObject> OnResponse;

        [JsonProperty]
        public Guid JobId { get; } = Guid.NewGuid();

        public void NotifyResponse(JObject resposne) =>
            OnResponse?.Invoke(this, resposne);
    }
}

File: src/XOPE Spy/dispatcher/type/impl/imessagewithresponse.cpp
#pragma once

#include "imessagewithresponse.h"


IMessageWithResponse::IMessageWithResponse(UiMessageType m) : IMessage(m), jobId(Guid::newGuid()) 
{
}

void IMessageWithResponse::serializeToJson(json& j)
{
	IMessage::serializeToJson(j);
	to_json(j, *this);
}

void to_json(json& j, IMessageWithResponse& mes)
{
	j["jobId"] = mes.jobId;
}